core/shared/src/main/scala/scalaz/zio/FiberRef.scala

```diff
@@ -0,0 +1,106 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package scalaz.zio
+
+import scalaz.zio
+
+/**
+ * Fiber's counterpart for Java's `ThreadLocal`. Value is automatically propagated
+ * to child on fork and merged back in after joining child.
+ * {{{
+ * for {
+ *   fiberRef <- FiberRef.make("Hello world!")
+ *   child <- fiberRef.set("Hi!).fork
+ *   result <- child.join
+ * } yield result
+ * }}}
+ *
+ * `result` will be equal to "Hi!" as changes done by child were merged on join.
+ *
+ * @param initial
+ * @tparam A
+ */
+final class FiberRef[A](private[zio] val initial: A) extends Serializable {
+
+  /**
+   * Reads the value associated with the current fiber. Returns initial value if
+   * no value was `set` or inherited from parent.
+   */
+  final val get: UIO[A] = modify(v => (v, v))
+
+  /**
+   * Sets the value associated with the current fiber.
+   */
+  final def set(value: A): UIO[Unit] = modify(_ => ((), value))
+
+  /**
+   * Returns an `IO` that runs with `value` bound to the current fiber.
+   *
+   * Guarantees that fiber data is properly restored via `bracket`.
+   */
+  final def locally[R, E, B](value: A)(use: ZIO[R, E, B]): ZIO[R, E, B] =
+    for {
+      oldValue <- get
+      b <- {
+        // TODO: Dotty doesn't infer this properly
+        val i0: ZIO.BracketAcquire_[R, E] = set(value).bracket_[R, E]
+        i0(set(oldValue))(use)
+      }
+    } yield b
+
+  /**
+   * Atomically modifies the `FiberRef` with the specified function.
+   */
+  final def update(f: A => A): UIO[A] = modify { v =>
+    val result = f(v)
+    (result, result)
+  }
+
+  /**
+   * Atomically modifies the `FiberRef` with the specified partial function.
+   * if the function is undefined in the current value it returns the old value without changing it.
+   */
+  final def updateSome(pf: PartialFunction[A, A]): UIO[A] = modify { v =>
+    val result = pf.applyOrElse[A, A](v, identity)
+    (result, result)
+  }
+
+  /**
+   * Atomically modifies the `FiberRef` with the specified function, which computes
+   * a return value for the modification. This is a more powerful version of
+   * `update`.
+   */
+  final def modify[B](f: A => (B, A)): UIO[B] = new ZIO.FiberRefModify(this, f)
+
+  /**
+   * Atomically modifies the `FiberRef` with the specified partial function, which computes
+   * a return value for the modification if the function is defined in the current value
+   * otherwise it returns a default value.
+   * This is a more powerful version of `updateSome`.
+   */
+  final def modifySome[B](default: B)(pf: PartialFunction[A, (B, A)]): UIO[B] = modify { v =>
```

Comment:
Love all these new methods, and by making them consistent, we're helping to make sure users don't have to re-learn a similar set of primitives for each new type of references (`Ref`, `RefM`, `TRef`, and `FiberRef`).